### ISSUE TITLE
Feature: Comments

### ISSUE THREAD

**Issue by RealOrangeOne** [hedgedoc/hedgedoc#657]
**Is your feature request related to a problem? Please describe.**

It'd be really useful to be able to comment on documents, whether logged in or not. Currently inlining comments into the text is a pain.

**Describe the solution you'd like**

The ability to select bits of text, and comment on them. Ideally with threading / suggestions.

This should be available regardless of whether the user is logged in. Perhaps prompting for a name when leaving a command if unauthenticated.

HackMD seem to do this feature quite nicely.

**Describe alternatives you've considered**

A full chat system is an option, but IMO completely unnecessary. If you want to chat, use a chat system. Comments are more for inline amendments / changes, and may not be made whilst the document is being edited.

**Additional context**

As mentioned in https://github.com/hedgedoc/hedgedoc/issues/554#issuecomment-721368767.

**Comment by DerMolly** [comment]
This issue seems to be in the same vain as https://github.com/hedgedoc/react-client/issues/529.

I'd suggest closing the issue here and adding a comment to that issue to keep all the ideas about comments in one place.

**Comment by RealOrangeOne** [comment]
Yup definitely looks the same to me! Hadn't spotted that some things were being tracked in the other repo. Will move thoughts.

**Comment by xeruf** [comment]
No, it isn't! In the other issue it was clearly stated that it does not cover that use-case:

>  This issue is mostly about adding criticmarkup. What you describe here ("comments/suggestions without changing the document") is another feature request. Please open another feature request issue for that.

**Comment by xeruf** [comment]
To supplement: There should be an additional in-between comment permission, like in Notion or Google Docs.
When one has no edit permissions or explicitly chooses comment mode, they can still edit the document, but their edit will turn into an interactive suggestion - a type of comment when it comes to handling in the backend, but with a UI as snappy as in Google Docs.

With that, I could finally move people over from Docs :)

**Comment by davidmehren** [comment]
I'll quote @mrdrogdrog from #2879:

> "I want to add comments to the rendered document, but the markdown code should not change and the comment should not appear in the code": This won't happen. The core dev team had a talk and we decided that we're strictly against meta data that are not part of the markdown code. Not only would it be hard to keep this data in sync with the markdown, they would also be lost when you export the note. The markdown code is the single source of truth. If we implement comments/suggestions then we'll use extensions like CriticMarkup.

**Comment by xeruf** [comment]
is there no way we can implement this in a way that can replace Docs?

**Comment by pirate** [comment]
Bumping this again. I love HedgeDoc but unfortunately our org will likely have to switch off HedgeDoc to HackMD if there's no openness to HedgeDoc adding this in the future.

**Comment by mrdrogdrog** [comment]
What you want are invisible comments, right? This is still possible using criticmarkup. E.g. by using a yet-to-be-built codemirror extension that hides the actual markup, marks the text in the UI and provides buttons to mark/comment text phrases.

We're not against comments as a UI feature, so I'm gonna reopen the issue to track the status of the user facing side of this feature.
We still have to decide on how to implement this.

**Comment by zeigerpuppy** [comment]
Hi @agger-magenta 
In our company, we're looking at ways to fund some development regarding this next year.
If your company has some resources for this too, we may be able to collaborate on getting this feature moving forward.

**Comment by bbuchalter** [comment]
### Preface

This proposal and design spec were written with the help of Claude (AI), but the decisions, trade-offs, and review are mine. I know open source maintainers are dealing with a flood of low-effort AI-generated issues and PRs right now, and I don't want to add to that burden. I'm [Brian Buchalter](https://www.linkedin.com/in/brianbuchalter/), a staff engineer, and I'm committed to doing the real work here — engaging with feedback, iterating on the design, and delivering code that meets this project's standards. Commenting is a feature I need for my own use of HedgeDoc, and one the community has been asking for since 2020. I'd like to help make it happen.

## Proposal: Inline Comments via CriticMarkup + CodeMirror 6 Extension

After reading through the discussion here, in #2879, #5450, and @ErikMichelson's [Matrix message](https://app.element.io/#/room/#hedgedoc:matrix.org/$04XXVHjQfsUuBiXzSQaQsboXdPRKe9CnuhZ8ZhwSx3w), I'd like to propose a concrete design that bridges the two camps: use CriticMarkup's highlight+comment syntax (`{== text ==}{>> comment <<}`) as the storage format — keeping markdown as the single source of truth — but build a CodeMirror 6 extension that hides the raw syntax and provides a Google Docs-style commenting UX with a sidebar panel. Users who prefer raw markdown see valid CriticMarkup. Users who want a clean experience see highlighted text and a comment sidebar. Same document, two views.

I understand that 1.x is in a stability-focused phase. This proposal is designed to respect that: the feature is **opt-in behind a config flag** (`HD_ENABLE_COMMENTS`, default `false`), requires **no backend logic changes**, **no database migrations**, and **no new API endpoints**. When disabled, it has zero effect on existing behavior. This is a pure frontend implementation — Yjs handles real-time sync for free since comments are just regular document edits.

The [full design spec](https://gist.github.com/bbuchalter/9f0d1f2cfe339e43b50d0468535790ab) covers the CriticMarkup threading format, three-panel UI layout, highlight styling, CM6 extension architecture, and edge cases. I'd welcome feedback from the maintainers before starting implementation — particularly on whether the threading convention within CriticMarkup is acceptable, and whether the three-panel layout approach works with where the project is headed.

## Mockup

<img width="1512" height="982" alt="Image" src="https://github.com/user-attachments/assets/58cbf72f-4d5c-4fe3-9fde-11f4d4ae8764" />

**Comment by bbuchalter** [comment]
Here is a simple proof of concept: https://github.com/bbuchalter/hedgedoc/pull/1

And a live demo server with commenting enabled: https://hedgedoc-comments-preview.fly.dev/n/91ce8s26zme0833s9s4jn7240r

**Comment by mrdrogdrog** [comment]
@bbuchalter 

You mixed something up here. 

1.x is the old stable version on the master branch. It doesn't use cm6 and won't receive new features. Not even with a feature flag. This is the version deployed on demo.hedgedoc.org

2.x is the in-develop version. This is the code on develop branch. This is the one using cm6. There is no need for a feature flag.

**Comment by bbuchalter** [comment]
Ah, yes. I realized that once I got into development that my [original comment](https://github.com/hedgedoc/hedgedoc/issues/657#issuecomment-4185972802) and the plan wasn't quite right about that. 👍

**Comment by mogul** [comment]
This looks super-cool, and I really appreciate balancing usability with the CriticMarkup inline standard! We are considering various proprietary options (eg TipTap with a bunch of paid add-ons) and this would make that unnecessary.

**Comment by bbuchalter** [comment]
@mrdrogdrog - it's been a few weeks, have you had a chance to look at the proof of concept in https://github.com/hedgedoc/hedgedoc/issues/657#issuecomment-4186668902

**Comment by zeigerpuppy** [comment]
Thanks @bbuchalter - this looks really good.

I like the way you have implemented the UI and 3-pane layout and the commenting functionality looks great.

We're also looking at spending some time on this feature over the next few months.

I think we may want to implement the storage of comments a little differently (as we need to keep the markdown exports compatible with our rmarkdown/quarto pipelines).

To that end, I think we'll be looking at storing the comments in a database and including just the short-UUID of the comment in the text (wrapped by an html comment).  

Currently the implementation you've shown edits the markdown as follows:

```md
do {== comments ==}{>>

---
@anonymous [2026-04-04T07:25Z]: they do ineed<<} work?
```

the alternative which we're looking at is:

```md
do <!--CStart-0TSoZbUSItGH--> comments <!--CEend-0TSoZbUSItGH--> work?
```

The main reason we're looking at that implementation is that it will allow all the metadata to be looked up for a particular comment (replies, dates, accept/reject comment, tags, comment-type) while not cluttering the markdown or making it difficult for interpreters without criticmark support.  I think it will also allow for nested comments and suggested edits.  We'd like to include the other criticmark elements too (deletions, additions, substitutions, highlighting).  

Potentially, there could be 2 hedgedoc markdown export options:

1. with your format (inline criticmark)
2. with only HTML comments (perhaps with a separate file exporting the comments too)

But I think it will be really good to work using a common front-end approach, as I can definitely predict that some users will prefer one format over the other!
